modules/social_features/social_core/src/Plugin/Field/FieldType/ExternalIdentifierItem.php

```diff
@@ -0,0 +1,221 @@
+      ]);
+
+    }
+
+    return $summary;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function schema(FieldStorageDefinitionInterface $field_definition): array {
+    return [
+      'columns' => [
+        'external_id' => [
+          'type' => 'varchar',
+          'description' => 'External ID',
+          'length' => 255,
+          'not null' => TRUE,
+        ],
+        'external_owner_target_type' => [
+          'type' => 'varchar',
+          'description' => 'External Owner Target Type',
+          'length' => 255,
+          'not null' => TRUE,
+        ],
+        'external_owner_id' => [
+          'type' => 'int',
+          'description' => 'External Owner ID',
+          'unsigned' => TRUE,
+          'not null' => TRUE,
+        ],
+      ],
+      'indexes' => [
+        'external_id' => ['external_id'],
+        'external_owner_target_type' => ['external_owner_target_type'],
+        'external_owner_id' => ['external_owner_id'],
+      ],
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setValue($values, $notify = TRUE): void {
+    if (is_array($values)) {
+      $this->set('external_id', $values['external_id'] ?? NULL);
+      $this->set('external_owner_target_type', $values['external_owner_target_type'] ?? NULL);
+      $this->set('external_owner_id', $values['external_owner_id'] ?? NULL);
+    }
+    else {
+      parent::setValue($values, $notify);
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function defaultStorageSettings() {
+    return [
+      'target_types' => [],
+    ] + parent::defaultStorageSettings();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function storageSettingsForm(array &$form, FormStateInterface $form_state, $has_data) {
+    $element['target_types'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Allowed external owner target types'),
+      '#description' => $this->t('Entity type of item to reference (Example: consumer)'),
+      '#default_value' => $this->getSetting('target_types'),
+      '#multiple' => TRUE,
+      // For field to work properly, at least one target type must be available,
+      // as this is hard requirement, If requirement can not be fulfilled, field
+      // will return an error if non-empty value will try to be applied.
+      // Despite all those facts, we still keep this option as optional, because
+      // by default, no target types are predefined.
+      '#required' => FALSE,
+      '#disabled' => $has_data,
+      '#size' => 10,
+    ];
+
+    // Only allow the field to target entity types that have an ID key. This
+    // is enforced in ::propertyDefinitions().
+    $entity_type_manager = \Drupal::entityTypeManager();
+    $filter = function (string $entity_type_id) use ($entity_type_manager): bool {
+      /** @var \Drupal\Core\Entity\EntityTypeInterface $entity_type */
+      $entity_type = $entity_type_manager->getDefinition($entity_type_id);
```

Comment:
Prefer using an `assert` statement which will fail in tests and more clearly demonstrate that if this doesn't happen there's a violation of base assumptions about the library we're using.

```suggestion
      /** @var \Drupal\Core\Entity\EntityTypeInterface $entity_type */
      $entity_type = $entity_type_manager->getDefinition($entity_type_id);
      assert($entity_type instanceof EntityTypeInterface, "EntitTypeManager::getDefinition should throw an exception if the entity type does not exist rather than returning NULL");
```

Reply:
I applied your suggestion and fix the small typo. Marking as resolved.